packages/agent/src/tool/user-input/UserInputTool.ts

```diff
@@ -0,0 +1,63 @@
+import zod from "zod";
+import { createToolAction } from "../ToolAction";
+import { ExecuteToolFunction } from "../ExecuteToolFunction";
+import { FormatResultFunction } from "../../action";
+import readline from "readline";
+
+export type UserInputInput = {
+  query: string;
+};
+
+export type UserInputOutput = {
+  text: string;
+};
+
+export const getUserInput = ({
+  id = "ask-user",
+  description = "Ask the user for input or to take an action.",
+  inputExample = {
+    query: "{question or action description}",
+  },
+  execute,
+  formatResult = ({ input, output: { text } }) => `${input.query}: ${text}`,
+}: {
+  id?: string;
+  description?: string;
+  inputExample?: UserInputInput;
+  execute: ExecuteToolFunction<UserInputInput, UserInputOutput>;
+  formatResult?: FormatResultFunction<UserInputInput, UserInputOutput>;
+}) =>
+  createToolAction({
+    id,
+    description,
+    inputSchema: zod.object({
+      query: zod.string(),
+    }),
+    outputSchema: zod.object({
+      text: zod.string(),
+    }),
+    inputExample,
+    execute,
+    formatResult,
+  });
+
+export const executeUserInput =
+  (): ExecuteToolFunction<UserInputInput, UserInputOutput> =>
+  async ({ input: { query } }) => {
+    const rl = readline.createInterface({
+      input: process.stdin,
+      output: process.stdout,
+    });
+
+    const text = await new Promise<string>((resolve) => {
+      rl.question(query, (answer) => {
+        resolve(answer);
+        rl.close();
+      });
+    });
+
+    return {
+      summary: `User input: ${text}`,
+      output: { text },
+    };
```

Comment:
Currently the user input tool is tied to the command line. I think this is fine for now, but eventually I want to think about decoupling, so it can be used in other environments. For example, when running the agent as a server process, it could be paused until the input is provided via e.g. a REST call. Actions don't have to be tied to tools - so we could have a user input action that does something that is not a tool execution (e.g. pause the agent and set it to a special user input state, or invoke a callback, etc.). I still need to think through the details - this is a good starting point though!

Reply:
yep agreed, for now I'm just using this to get into your concepts and understand your way of thinking about these things, plus as a starting point for many users experimenting with it it'll be fine. we can still extend it later.